### PR TITLE
lmao glsl is hot garbage

### PR DESCRIPTION
**Pull request by Tossrock**
this does render a waveform with a shader tho
image

unfortunately i am not smart enough to Do A React Fiber MobX UiState so it doesn't respond to scroll events or nothin'

### CODE REVIEW

**Comment by Tossrock** [comment]
Did I spend like an hour desperately trying to figure out why tf my shader wasn't doing anything before realizing I hadn't declared the params as `uniform`? Maybe. Maybe.

File: src/components/Waveform.tsx [hunk 1]
```diff
@@ -2,73 +2,134 @@ import { useStore } from "@/src/types/StoreContext";
...
-        hideScrollbar: true,
-        interact: false,
-        minPxPerSec: INITIAL_PIXELS_PER_SECOND,
-      };
-      wavesurferRef.current = WaveSurfer.create(options);
-      wavesurferRef.current.zoom(INITIAL_PIXELS_PER_SECOND);
+          const sampleCount = audioBuffer.length;
+          const sampleTexWidth = 1024;
+          const sampleTexHeight = Math.floor(sampleCount / (4 * sampleTexWidth));
+          const sampleTex = new THREE.DataTexture(floatArray, sampleTexWidth, sampleTexHeight, THREE.RGBAFormat, THREE.FloatType);
```
Comment (Tossrock) [inline src/components/Waveform.tsx:51]: Unfortunately, this is the only way to pass the required sample data to the shader. Using a plain float array doesn't work for several reasons:

1. Arrays can't be dynamically sized. You can have a `float[16] samples`, but you can't have a `float[] samples`.

Now, you are no doubt thinking, "ah but can't we just preallocate a very large array, large enough for any possible song, and then pass in the number of samples to read from it?" And unfortunately,

2. Arrays can't be very big. `float[[1024]` is about what you're working with. The Cloudkicker song has ~15M samples.

So, data textures are the only real solution. Even those are pushing it, the Cloudkicker track ends up being 1024 x 3900, which is pushing 4k texture size, a common HW limit. Some very modern HW supports 8k+ texture resolution, but, you know, #compatibility.

**Comment by brollin** [comment]
Rebased, deconflicted, and simplified a bit. I think I will convert to R3F next!

File: src/components/Waveform.tsx (hunk 1 above)
Comment (brollin) [reply src/components/Waveform.tsx]: > Now, you are no doubt thinking, "ah but can't we just preallocate a very large array, large enough for any possible song, and then pass in the number of samples to read from it?"

I was thinking that, get out of my head!

**Comment by brollin** [comment]
Ok, just found something. Apparently the wavesurfer guy is actively developing v7 which fixes the performance issues with zooming. He only started 2 weeks ago, but the alpha is available and trying it out, it zooms quite well. There are some styling issues, but they seem fixable.  

https://github.com/katspaugh/wavesurfer.ts

Since that library is quite featureful and now actually performant, and we've got it all wired up already, I'm tempted to just move ahead with that. Trying to race to the MVP asap!

I'd say a future with this kind of shader-based waveform could be desirable for maximum customization. We would have to complete the zooming stuff and add in audio playback/seeking. Optionally coloring the waveform up to the point that has played as well.

I will create a simple toggle to switch between waveform renderers for now, so both can be worked on I'm thinking.

**Comment by brollin** [comment]
image

Ok, both exist now. In the above screenshot how there is a button force switching between them. With the new wavesurfer library zooming is now performant, but there is some bug with the pixelsPrSecond, so there's an annoying offset. Will try to fix that later/file an issue.

**Comment by Tossrock** [comment]
> Ok, just found something. Apparently the wavesurfer guy is actively developing v7 which fixes the performance issues with zooming. He only started 2 weeks ago, but the alpha is available and trying it out, it zooms quite well. There are some styling issues, but they seem fixable.
> 
> https://github.com/katspaugh/wavesurfer.ts
> 
> Since that library is quite featureful and now actually performant, and we've got it all wired up already, I'm tempted to just move ahead with that. Trying to race to the MVP asap!
> 
> I'd say a future with this kind of shader-based waveform could be desirable for maximum customization. We would have to complete the zooming stuff and add in audio playback/seeking. Optionally coloring the waveform up to the point that has played as well.
> 
> I will create a simple toggle to switch between waveform renderers for now, so both can be worked on I'm thinking.

Oh nice! I mean, if the existing integrated library solved their performance problem, I'd just stick with that?

**Comment by brollin** [comment]
Maybe down the road we will cut out the shader waveform, but it's nice to have around for now. The new library actually has several bugs, clearly in the alpha stages. About to file a second issue. I'm thinking that how quickly they can get to those could determine whether we stick with it or build our own.

**Pull request merged**